src/optimizer/input_column_deriver.cpp

```diff
@@ -96,6 +101,61 @@ void InputColumnDeriver::Visit(const Limit *op) {
   output_input_cols_ = std::make_pair(std::move(cols), std::move(child_cols));
 }
 
+void InputColumnDeriver::Visit(const CteScan *op) {
+  // All aggregate expressions and TVEs in the required columns and internal
+  // sort columns are needed by the child node
+  ExprSet input_cols_set;
+  for (auto expr : required_cols_) {
+    if (parser::ExpressionUtil::IsAggregateExpression(expr)) {
+      input_cols_set.insert(expr);
+    } else {
+      parser::ExpressionUtil::GetTupleValueExprs(&input_cols_set, expr);
+    }
+  }
+
+  std::vector<common::ManagedPointer<parser::AbstractExpression>> cols;
+  for (const auto &expr : input_cols_set) {
+    cols.push_back(expr);
+  }
+
+  PT2 child_cols;
+
+  child_cols.reserve(gexpr_->GetChildrenGroupsSize());
+  for (size_t i = 0; i < gexpr_->GetChildrenGroupsSize(); i++) {
+    auto child_exprs = op->GetChildExpressions()[i];
+    std::vector<common::ManagedPointer<parser::AbstractExpression>> new_child_exprs;
+    new_child_exprs.reserve(child_exprs.size());
+    for (auto &elem : child_exprs) {
+      new_child_exprs.push_back(elem);
+    }
+    child_exprs.clear();
+    child_exprs = new_child_exprs;
+    child_cols.push_back(std::move(child_exprs));
```

Comment:
Can there be a comment included here explaining why this logic is necessary/correct?